cmake/modules/FindDEAL_II_KOKKOS.cmake

```diff
@@ -19,37 +19,92 @@
 #   KOKKOS_INTERFACE_LINK_FLAGS
+#   KOKKOS_VERSION
 #
 
 set(KOKKOS_DIR "" CACHE PATH "An optional hint to a Kokkos installation")
 set_if_empty(KOKKOS_DIR "$ENV{KOKKOS_DIR}")
 
+# silence a warning when including FindKOKKOS.cmake
+set(CMAKE_CXX_EXTENSIONS OFF)
 
-if(DEAL_II_TRILINOS_WITH_KOKKOS OR DEAL_II_PETSC_WITH_KOKKOS)
-  # Let ArborX know that we have found Kokkos
-  set(Kokkos_FOUND ON)
-  # Let deal.II know that we have found Kokkos
-  set(KOKKOS_FOUND ON)
-else()
-  # silence a warning when including FindKOKKOS.cmake
-  set(CMAKE_CXX_EXTENSIONS OFF)
-  find_package(Kokkos 3.7.0 QUIET
-    HINTS ${KOKKOS_DIR} ${Kokkos_DIR} $ENV{Kokkos_DIR}
+# The minimal Kokkos version we support
+set(KOKKOS_MINIMUM_REQUIRED_VERSION 3.7.0)
+
+#
+# Make sure that we prioritize Kokkos bundled with Trilinos or PETSc.
+#
+# We do this by searching for the Kokkos installation first at the
+# restrictive path locations we deduced from both configurations.
+#
+
+if(Kokkos_FOUND)
+  message(WARNING "\n"
+    "find_package(Kokkos [...]) already called. The deal.II CMake "
+    "configuration might be inconsistent.\n\n"
     )
+endif()
 
-  set(KOKKOS_FOUND ${Kokkos_FOUND})
+if(TRILINOS_WITH_KOKKOS)
+  message(STATUS "Found Trilinos with bundled Kokkos library. Overriding search path.")
+  set(KOKKOS_MINIMUM_REQUIRED_VERSION 3.4.0)
+  find_package(Kokkos ${KOKKOS_MINIMUM_REQUIRED_VERSION} QUIET
+    PATHS ${TRILINOS_KOKKOS_DIR} NO_DEFAULT_PATH
+    )
 
-  set(_target Kokkos::kokkos)
-  process_feature(KOKKOS
-    TARGETS REQUIRED _target
+  if(NOT Kokkos_FOUND)
+    message(WARNING "\n"
+      "The find_package(Kokkos [...]) call failed with the path deduced from "
+      "the Trilinos configuration. The deal.II CMake configuration might be "
+      "inconsistent.\n\n"
+      )
+  endif()
+
+elseif(PETSC_WITH_KOKKOS)
+  message(STATUS "Found PETSc with bundled Kokkos library. Overriding search path.")
+  set(KOKKOS_MINIMUM_REQUIRED_VERSION 3.4.0)
+  find_package(Kokkos ${KOKKOS_MINIMUM_REQUIRED_VERSION} QUIET
+    PATHS ${PETSC_KOKKOS_DIR} NO_DEFAULT_PATH
     )
+
+  if(NOT Kokkos_FOUND)
+    message(WARNING "\n"
+      "The find_package(Kokkos [...]) call failed with the path deduced from "
+      "the PETSc configuration. The deal.II CMake configuration might be "
+      "inconsistent.\n\n"
+      )
+  endif()
 endif()
 
+#
+# Find the Kokkos CMake configuration:
+#
+
+find_package(Kokkos ${KOKKOS_MINIMUM_REQUIRED_VERSION} QUIET
+  HINTS ${KOKKOS_DIR} ${Kokkos_DIR} $ENV{Kokkos_DIR}
+  )
 
+set(_target)
 if(Kokkos_FOUND)
+  set(_target Kokkos::kokkos)
+endif()
+
+process_feature(KOKKOS
+  TARGETS REQUIRED _target
+  CLEAR Kokkos_DIR
+  )
+
+if(KOKKOS_FOUND)
+  #
+  # Set version number:
+  #
   if(NOT Kokkos_VERSION)
-    message(FATAL_ERROR "FindPackage(Kokkos) did not set Kokkos_VERSION!")
+    message(WARNING "\n"
+      "find_package(Kokkos [...]) did not set Kokkos_VERSION!"
+      )
+    set(Kokkos_VERSION ${KOKKOS_MINIMUM_REQUIRED_VERSION})
```

Comment:
@bangerth Let us base https://github.com/dealii/dealii/pull/18220 on top of the refactoring done in this pull request. Specifically: All you'd need to do is change the `WARNING` to a `STATUS` and extract the version number from the header file as a backup.